### DISCUSSION TITLE
Scalability - Design discussion

### DISCUSSION THREAD

**Discussion by lucasponce** [kiali/kiali#4080] [Roadmap]
Hi everyone,

As defined in https://github.com/kiali/kiali/discussions/4063 Scalability is one active area in the Kiali project.

The goal of a Service Mesh is to build a highly scalable service to service architecture and hence Kiali’s goal is to provide smoothly visualization in scenarios with potentially multiple clusters, hundreds of namespaces and thousands of workloads.

On highly scalable scenarios there are some aspects from user perspective to consider in the design:

- How to visualize large scenarios in the Graph: expand/collapse per namespace, cluster and/or other criteria.

- How to filter (a.k.a. “find/hide)) the information interesting for the user: based on “graph queries” (show part of the graph with status == X), introduce more flexible ways to filter like using user’s labels or annotations.

- How to navigate: on large scenarios probably focus and zoom should be also driven by Kiali importing GIS vectorial techniques (i.e. I start visualizing the whole architecture, and after one query, Kiali may zoom in and refocus the center of the user in the interested elements).

- Also, after introducing more filtering capabilities, filtering status should be consistent across all pages of Kiali (I perform a filter in the Graph based on labels, then the same filter state should be used in other pages/elements until the user changes it).

On the implementation side, the Kiali architecture should improve its design to support these highly scalable scenarios, some aspects to consider:

- Use cache techniques with the information required to build the different information that Kiali processes (in a nutshell, Kiali queries telemetry, tracing, kubernetes configuration, istio configuration and istio registry to compose all the data). There are no silver bullets in this design, caching has a cost and a tradeoff should be discussed. Today the caching timeout is not visible to the user, perhaps in highly scalable scenarios, user should be able to change the validity of the info cached (i.e. last query from 5 minutes ago, etc).

- To enable advanced filtering, some features should be composed in the backend instead in the frontend: health calculation, config validations are just two examples that are calculated in a separated backend endpoints and put together in the UI. If now we may want to improve the filtering and/or introduce new capabilities (like filtering by user labels/annotations) the backend should support a composition of all of these responses (and the previous caching point is important to be able to respond smoothly to these changes).

- Deployment models (multicluster/federation scenarios) should be considered but note those will be discussed in a separate discussion.

With this context, the aim of this discussion is to openly brainstorm about these topics in two directions:
- From a user perspective, is there any feature/capability that would be good to focus/prioritize ?
- From an implementation perspective, which design improvements would help Kiali to address all of these scalability challenges.

Please, join us in these discussions :).

**Comment by nrfox** [comment]
Another area of focus for scalability could be frontend performance. Things like how long it takes to render the graph or charts, how does the size of the graph affect render times, system resources required to render large graphs. Ideally the UI would feel responsive and fast without consuming all of your system's resources even with a large mesh. If the UI begins to feel slow or clunky imo this has a real negative impact on user experience.

**Reply by lucasponce** [reply]
Yes, even I think probably this may trigger more aspects in the backend rather in the frontend.
I mean, typically, to render "thousands" of nodes is not a problem for todays browser.
I like this benchmark [1] as even it's a very different domain, it has some aspects that I like:
- it has thousands of nodes to show
- it allows to performs queries to extract information
- the relations between nodes are pretty unstructured which may or may not represent microservices topologies

So, in terms of pure UI, it sounds doable, I guess focus on scalability is how to generate this information in a efficient way.

In that sense, one idea is that probably we may want to revisit some logic and decide where to place it (frontend / backend) and to introduce better use cases for caching.

But this topic is also pretty open as well, but in my thoughts the goal would be to move into a model that is more asynchornous and perhaps can be updated more often but without creating long "stops" to wait everything is being computed.

[1] https://js.cytoscape.org/demos/tokyo-railways/

**Comment by lucasponce** [comment]
Some of the ideas to consider in the design of the Graph model:

- On highly populated scenarios, allow Kiali to present a Graph of namespace connections (not including workloads until user may zoom in / drill down a namespace).

- On large scenarios, users may define some "filter" about the relations they want to visualize:
"As owner of Service A, deployed on namespace A, I'm interested to see how my service works with others namespaces that are related, these namespaces may be directly connected to Service A or indirectly; but I could define some criteria to see only this partial view in the Graph (as a way to reduce the space to visualize)".

- "Filtering" capabilities should be strong and flexible, then the design of the Graph model should support to "compose" other information that today is managed separately (health info, validations, additional info collected from kubernetes model as labels/annotations that can be used as filter criteria).

These ideas will require changes in how a graph model is calculated, adding more caching challenges; so probably Kiali would miss more "real-time" capabilities in favor of "support big deployments" in a efficient way.

**Reply by xeviknal** [reply]
Related to filtering, I'd like to add the ability of users to pick which are the default filtering I want to use. Wearing the had of a developer or devops, I only want to focus on the `app: user_points`. 

This might trigger to the point of managing users within kiali. User A sees the `app: user_points | users` and User B only sees `app: bets | matches`.

AFAIK, now we have one set of config per kiali instance. This is quite limiting.

**Reply by lucasponce** [reply]
I agree that there could be roles/groups that can drive some configuration personalization.
This personalization shouldn't be "critical" to not introduce potential new elements in the architecture with a high cost to maintain.
(I mean, if for some reason, the personalization is gone, the defaults could be just ok).

**Comment by xeviknal** [comment]
I really liked the idea of doing rankings of unhealthy resources to allow users to navigate through kiali driven by unhealthy resources. e.g.: Top 3 most unhealthier clusters -> Cluster A is not healthy (click) -> top 3 most unhealthier namespaces -> Namespace B is not healthy (click) -> top 3 most unhealthier apps/workloads/requests -> App C is not healthy (click) -> Metrics for App C.
Kiali might benefit from a more clearer (on huge archs) navigation to troubleshoot what's wrong.

**Reply by abonas** [reply]
How app health is related to scalability?

**Reply by xeviknal** [reply]
Didn't see the replay. Tons of them.
It's related to the navigation. Now we have the navigation flow willing to cover everything: all namespaces, all workloads, all istio objects. And this doesn't scale because you might have hundreds/thousands of them. Therefore, changing the navigation approach from this to showing the most N problematic/unhealthy items (apps, objects, workloads, etc). This way, you focus on what you have to put your hands on instead of seeing everything: the important and the no important with same level of importance.

Rankings are a good way to define which are the top most healthy/unhealthy/valid items.

**Comment by xeviknal** [comment]
**Scalability for validations** Kiali should have a robust model supporting the validation feature for big big clusters. As of today, the validation endpoints have this poor performance: 

> On a cluster with 1600 services and ~2000pods the tls and validations api endpoints take a long time to respond in large environment. For example tls takes 1.4 minutes to return.

> Valiation endpoints result: {"errors":1,"objectCount":536,"warnings":73}. Load time is 30 seconds .

from https://github.com/kiali/kiali/issues/4224

The first mechanism that comes to my mind is pre-processing or caching the validation status for each object in the mesh. The expiration date for each object should be when there is one new istio object related to the validation affecting to that object. Let's see one example:

Gateway validations only need the other gateways and workloads:
```go
type GatewayChecker struct {
	GatewaysPerNamespace  [][]kubernetes.IstioObject
	Namespace             string
	WorkloadsPerNamespace map[string]models.WorkloadList
}
```

Therefore the cache for all the gateway objects should expire when there is either one new workload or gateway. Otherwise, they are still valid.

For example, sidecars validations:

```go
type SidecarChecker struct {
	Sidecars       []kubernetes.IstioObject
	ServiceEntries []kubernetes.IstioObject
	Services       []core_v1.Service
	Namespaces     models.Namespaces
	WorkloadList   models.WorkloadList
}
```

The cache/pre-process can be expired or recalculated when there is a new sidecar, services entry, service, namespace or workload. If you add a DestinationRule, for instance, its validity remains the same.

**Reply by xeviknal** [reply]
Similar approach should be placed for **mTLS status**. It might be a bit expensive to compute tls for both mesh-level and ns-level.

**Reply by lucasponce** [reply]
About validation + scalability I have three ideas/thoughts pretty open to discuss:
- Revisit the validations that today Kiali implements and propose to move to Istio those that make sense (aiming to try to move most of them).
- Rely more in the "istioctl analyze" and "status" (when user enables them) to build a cache of validation results (the goal would be to move some of the load of the validation outside of Kiali, and just consume it, again, it's pretty open).
- Also, I'm open to revisit the implementation design, and see if a webhook design could help in some of the validations (but this is just an idea).

I mean, I'd like to centralize more the validations in a single point and let Kiali consume them rather than process them, and in a first thought the logical place is to try to move them to the istio logic.

I know it's open and a lot of details to discuss, but I guess that reflects my thoughts here as well.

**Comment by nrfox** [comment]
After testing a few setups in my local environment what I've noticed is:
- Adding new namespaces greatly increases response times for both the `/tls` and `/validations` endpoints. These are just empty namespaces with nothing in them.
- Setting `accessible_namespaces` to anything other than `**` increases response times for `/tls` and `/validations` endpoints.

You can test these various scenarios by doing the following and measuring response times at each step in your browser:
- Deploying bookinfo
- Adding a couple hundred DRs to bookinfo. (Can be the same rules with just the name changed)
- Creating 10s of empty namespaces
- Setting `accessible_namespaces` to something other than `**`

I was expecting adding DRs to modestly increase response times for the `/tls` and `/validations` endpoints but adding empty namespaces and setting `accessible_namespaces` was a surprise.

Note that you can test all of the above without needing a large cluster since it does not create any pods that require system resources. This might be true for other endpoints as well. Having a baseline set of benchmarks for the various endpoints will go a long way in understanding where the scalability issues lie and how to fix them.

**Comment by jshaughn** [comment]
Scalability is a big topic, and within that, **Graph Scalability** is a big topic.  I've added two breakout discussions to discuss a couple of specific proposals:

[Revisiting design of health on the graph](https://github.com/kiali/kiali/discussions/4507)
[Namespace Graphs](https://github.com/kiali/kiali/discussions/4508)
